src/QsCompiler/TextProcessor/QsTypeParsing.fs

```diff
@@ -171,73 +172,89 @@ let private operationType =
 
-    let opTypeWith characteristics =
-        let withInnerBrackets = optTupleBrackets (tupleBrackets inAndOutputType |>> fst .>>. characteristics)
-        let withoutInnerBrackets = optTupleBrackets (inAndOutputType .>>. characteristics)
-        withInnerBrackets <|> withoutInnerBrackets
-
-    let deprecatedCharacteristics =
-        let colonWithWarning = buildWarning (getEmptyRange .>> colon) WarningCode.DeprecatedOpCharacteristicsIntro
-
-        attempt (colonWithWarning >>. characteristics .>> notFollowedBy (comma >>. quantumFunctor))
-        <|> (qsCharacteristics.parse |>> (fun r -> r.Start) <|> (getPosition .>> colon) >>= functorSupport)
-
+    let opTypeWith characteristics = inAndOutputType .>>. characteristics
+    let opTypeWithoutCharacteristics = inAndOutputType .>>. preturn (Characteristics.New(EmptySet, Null))
+    let deprecatedCharacteristics = qsCharacteristics.parse |>> (fun r -> r.Start) >>= functorSupport
     let characteristics =
         qsCharacteristics.parse >>. expectedCharacteristics isTupleContinuation
         .>> notFollowedBy (comma >>. quantumFunctor)
 
-    let opTypeWithoutCharacteristics =
-        optTupleBrackets (inAndOutputType .>>. preturn ((EmptySet, Null) |> Characteristics.New))
-
-    opTypeWith characteristics <|> opTypeWith deprecatedCharacteristics <|> opTypeWithoutCharacteristics
-    |>> asType Operation // keep this order!
-
-/// Parses a Q# function type raising the corresponding missing bracket errors if the outer tuple brackets are missing.
-/// NOTE: Uses leftRecursionByInfix to process the signature and raise suitable errors.
+    // keep this order!
+    attempt (opTypeWith characteristics)
+    <|> attempt (opTypeWith deprecatedCharacteristics)
+    <|> opTypeWithoutCharacteristics
+    |> term
+    |>> asType Operation
+
+/// <summary>
+/// Parses a Q# function type.
+/// </summary>
+/// <remarks>
+/// Uses leftRecursionByInfix to process the signature and raise suitable errors.
+/// </remarks>
 let private functionType =
-    let core = leftRecursionByInfix fctArrow qsType (expectedQsType isTupleContinuation)
-    optTupleBrackets core |>> asType Function
+    leftRecursionByInfix fctArrow qsType (expectedQsType isTupleContinuation) |> term
+    |>> asType Function
 
 /// Parses a Q# tuple type, raising an Missing- or InvalidTypeDeclaration error for missing or invalid items.
 /// The tuple must consist of at least one tuple item.
 let internal tupleType =
     let buildTupleType (items, range: Range) = (TupleType items, range) |> QsType.New
     buildTuple qsType buildTupleType ErrorCode.InvalidType ErrorCode.MissingType invalidType
 
+/// <summary>
+/// Validates that a parsed Q# type follows additional rules that prevent confusing syntax.
+/// </summary>
+/// <param name="isArrayItem">Whether the given type is part of an array postfix bracket type, "T[]" or "T[n]".</param>
+let internal validateTypeSyntax isArrayItem { Type = kind; Range = range } =
```

Comment:
I see you went with disallowing this after all. :)

Reply:
Originally, I just disallowed `Qubit => Unit is Adj[]` as we discussed, but then I realized the same function could be used to check `new Int -> Int[3]` as well, so it seemed like a good idea.